src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolume.java

```diff
@@ -1,9 +1,409 @@
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-amz-meta-" + entry.getKey(), entry.getValue());
+            }
+
+            break;
+
+        case AZURE:
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-ms-meta-" + entry.getKey(), entry.getValue());
+            }
+            break;
+
+        case GCS:
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-goog-meta-" + entry.getKey(), entry.getValue());
+            }
+            break;
+    }
+    */
+  }
+
+  PresignedUrlInfo dequeueUrlInfo() {
+    PresignedUrlInfo info = this.presignedUrlInfos.poll();
+    boolean generate = false;
+    if (info == null) {
+      generate = true;
+    } else {
+      // Since the queue had a non-null entry, there is no way numUrlsInQueue is <=0
+      int remainingUrlsInQueue = this.numUrlsInQueue.decrementAndGet();
+      if (remainingUrlsInQueue <= LOW_WATERMARK_FOR_EARLY_REFRESH) {
+        generate = true;
+        // assert remaininUrlsInQueue >= 0
+      }
+    }
+    if (generate) {
+      // TODO: do this generation on a background thread to allow the current thread to make
+      // progress ?
+      // Will wait for perf runs to know this is an issue that needs addressal.
+      generateUrls(LOW_WATERMARK_FOR_EARLY_REFRESH);
+    }
+    return info;
+  }
+
+  // NOTE : We are intentionally NOT re-enqueuing unused URLs here as that can cause correctness
+  // issues by
+  // accidentally enqueuing a URL that was actually used to write data out. Its okay to allow an
+  // unused URL to go waste
+  // as we'll just go out and generate new URLs. Do NOT add an enqueueUrl() method for this reason.
+
+  private void generateUrls(int minCountToSkipGeneration) {
+    int numAcquireAttempts = 0;
+    boolean acquired = false;
+
+    while (!acquired && numAcquireAttempts++ < 300) {
+      // Use an aggressive timeout value as its possible that the other requests finished and added
+      // enough
+      // URLs to the queue. If we use a higher timeout value, this calling thread's flush is going
+      // to
+      // unnecessarily be blocked when URLs have already been added to the queue.
+      try {
+        acquired = this.generateUrlsSemaphore.tryAcquire(1, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        // if the thread was interrupted there's nothing we can do about it, definitely shouldn't
+        // continue processing.
+        String message =
+            String.format(
+                "Semaphore acquisition in ExternalVolume.generateUrls was interrupted, "
+                    + "likely because the process is shutting down. TableRef=%s",
+                tableRef);
+        logger.logError(message);
+        throw new SFException(ErrorCode.INTERNAL_ERROR, message);
```

Comment:
I think we should restore the interrupted flag here in case some higher level adds a retry.

Reply:
good catch! fixed.